File: packages/android/fernlink-sdk/src/main/kotlin/xyz/fernlink/sdk/wifi/TcpFraming.kt
package xyz.fernlink.sdk.wifi

import java.io.InputStream
import java.io.OutputStream

/**
 * Simple length-prefix framing for TCP streams.
 *
 * Wire format per message: [typeTag: 1 byte][length: 4 bytes BE][payload: length bytes]
 *
 * typeTag values:
 *   TYPE_REQUEST = 0x01 — VerificationRequest JSON
 *   TYPE_PROOF   = 0x02 — VerificationProof JSON
 */
internal object TcpFraming {

    const val TYPE_REQUEST: Byte = 0x01
    const val TYPE_PROOF:   Byte = 0x02

    /** Write one framed message to the output stream. Thread-safe per stream with external sync. */
    fun write(out: OutputStream, typeTag: Byte, payload: ByteArray) {
        val len = payload.size
        out.write(byteArrayOf(
            typeTag,
            (len shr 24 and 0xFF).toByte(),
            (len shr 16 and 0xFF).toByte(),
            (len shr  8 and 0xFF).toByte(),
            (len        and 0xFF).toByte(),
        ))
        out.write(payload)
        out.flush()
    }

    /**
     * Read one complete frame from the input stream. Blocks until the full
     * frame is received or throws if the stream is closed.
     *
     * Returns Pair(typeTag, payload) or null if the stream reached EOF cleanly.
     */
    fun read(input: InputStream): Pair<Byte, ByteArray>? {
        val header = ByteArray(5)
        var totalRead = 0
        while (totalRead < 5) {
            val n = input.read(header, totalRead, 5 - totalRead)
            if (n < 0) return null   // EOF
            totalRead += n
        }
        val typeTag = header[0]
        val length  = ((header[1].toInt() and 0xFF) shl 24) or
                      ((header[2].toInt() and 0xFF) shl 16) or
                      ((header[3].toInt() and 0xFF) shl  8) or
                       (header[4].toInt() and 0xFF)
        if (length <= 0 || length > 1_048_576) return null   // sanity cap 1 MB

        val payload = ByteArray(length)
        var read = 0
        while (read < length) {
            val n = input.read(payload, read, length - read)
            if (n < 0) return null
            read += n
        }
        return Pair(typeTag, payload)
    }
}

File: packages/ble-desktop/src/wifi/tcp_framing.rs
use std::io::{Read, Write};
use anyhow::{bail, Result};
use tokio::io::{AsyncReadExt, AsyncWriteExt};

pub const TYPE_REQUEST: u8 = 0x01;
pub const TYPE_PROOF:   u8 = 0x02;

/// Write one framed message: [typeTag: 1 byte][length: 4 bytes BE][payload]
pub fn write_frame(stream: &mut impl Write, type_tag: u8, payload: &[u8]) -> Result<()> {
    let len = payload.len() as u32;
    stream.write_all(&[type_tag])?;
    stream.write_all(&len.to_be_bytes())?;
    stream.write_all(payload)?;
    stream.flush()?;
    Ok(())
}

/// Read one complete frame. Blocks until the full frame is received.
/// Returns (typeTag, payload) or None on clean EOF.
pub fn read_frame(stream: &mut impl Read) -> Result<Option<(u8, Vec<u8>)>> {
    let mut header = [0u8; 5];
    match stream.read_exact(&mut header) {
        Ok(()) => {}
        Err(e) if e.kind() == std::io::ErrorKind::UnexpectedEof => return Ok(None),
        Err(e) => return Err(e.into()),
    }
    let type_tag = header[0];
    let length   = u32::from_be_bytes([header[1], header[2], header[3], header[4]]) as usize;
    if length > 1_048_576 { bail!("frame too large: {length}"); }

    let mut payload = vec![0u8; length];
    stream.read_exact(&mut payload)?;
    Ok(Some((type_tag, payload)))
}

// ── Async variants (Tokio) ────────────────────────────────────────────────────

pub async fn write_frame_async(
    stream: &mut (impl AsyncWriteExt + Unpin),
    type_tag: u8,
    payload: &[u8],
) -> Result<()> {
    let len = payload.len() as u32;
    stream.write_all(&[type_tag]).await?;
    stream.write_all(&len.to_be_bytes()).await?;
    stream.write_all(payload).await?;
    stream.flush().await?;
    Ok(())
}

pub async fn read_frame_async(
    stream: &mut (impl AsyncReadExt + Unpin),
) -> Result<Option<(u8, Vec<u8>)>> {
    let mut header = [0u8; 5];
    match stream.read_exact(&mut header).await {
        Ok(_)  => {}
        Err(e) if e.kind() == std::io::ErrorKind::UnexpectedEof => return Ok(None),
        Err(e) => return Err(e.into()),
    }
    let type_tag = header[0];
    let length   = u32::from_be_bytes([header[1], header[2], header[3], header[4]]) as usize;
    if length > 1_048_576 { bail!("frame too large: {length}"); }
    let mut payload = vec![0u8; length];
    stream.read_exact(&mut payload).await?;
    Ok(Some((type_tag, payload)))
}
